src/jedi_nft.cairo

```diff
@@ -0,0 +1,459 @@
+    rank: u128,
+    score: u128,
+    level: u8,
+    total_eligible_users: u128,
+}
+
+#[starknet::interface]
+trait IJediNFT<TContractState> {
+    fn tokenURI(self: @TContractState, token_id: u256) -> Span<felt252>;
+
+    fn contractURI(self: @TContractState) -> Span<felt252>;
+
+    fn is_completed(self: @TContractState, task_id: u128, address: ContractAddress) -> bool;
+
+    fn set_merkle_root(ref self: TContractState, task_id: u128, merkle_root: felt252);
+
+    fn get_merkle_root(self: @TContractState, task_id: u128) -> felt252;
+
+    fn set_mint_sig_pub_key(ref self: TContractState, mint_sig_public_key: felt252);
+
+    fn get_mint_sig_pub_key(self: @TContractState) -> felt252;
+
+    fn mint_sig(
+        ref self: TContractState,
+        token_id: u128,
+        signature: Span<felt252>,
+        token_metadata: TokenMetadata
+    );
+
+    fn mint_whitelist(
+        ref self: TContractState,
+        token_id: u128,
+        proof: Array<felt252>,
+        token_metadata: TokenMetadata
+    );
+
+    fn get_token_metadata(self: @TContractState, token_id: u128) -> TokenMetadata;
+}
+
+#[starknet::contract]
+mod JediNFT {
+    use option::OptionTrait;
+    use traits::{Into, TryInto, Default, Felt252DictValue};
+    use array::{SpanSerde, ArrayTrait};
+    use clone::Clone;
+    use array::SpanTrait;
+    use box::BoxTrait;
+    use ecdsa::check_ecdsa_signature;
+    use zeroable::Zeroable;
+    use openzeppelin::token::erc721::ERC721;
+    use openzeppelin::token::erc721::ERC721::InternalTrait as ERC721InternalTrait;
+    use openzeppelin::introspection::interface::ISRC5;
+    use openzeppelin::introspection::interface::ISRC5Camel;
+    use openzeppelin::token::erc721::interface::{
+        IERC721, IERC721CamelOnly, IERC721Metadata, IERC721MetadataCamelOnly
+    };
+    use jedinft::access::ownable::{Ownable, IOwnable};
+    use jedinft::access::ownable::Ownable::{
+        ModifierTrait as OwnableModifierTrait, InternalTrait as OwnableInternalTrait,
+    };
+    use starknet::ContractAddress;
+    use starknet::get_caller_address;
+    use alexandria_data_structures::merkle_tree::{
+        Hasher, MerkleTree, pedersen::PedersenHasherImpl, MerkleTreeTrait, MerkleTreeImpl
+    };
+    use jedinft::storage::StoreSpanFelt252;
+    use super::TokenMetadata;
+
+    const IERC721_ID_LEGACY: felt252 = 0x80ac58cd;
+    const IERC721_METADATA_ID_LEGACY: felt252 = 0x5b5e139f;
+    const IERC721_RECEIVER_ID_LEGACY: felt252 = 0x150b7a02;
+
+
+    #[storage]
+    struct Storage {
+        _completed_tasks: LegacyMap::<(u128, ContractAddress), bool>,
+        _merkle_roots: LegacyMap::<u128, felt252>,
+        _uri: Span<felt252>,
+        _contract_uri: Span<felt252>,
+        _mint_sig_public_key: felt252,
+        // (token_id) -> (task_id, name, rank, score, level, total_eligible_users)
+        _token_metadata: LegacyMap::<u128, TokenMetadata>,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
```

Comment:
Will these event objects be used? Right now they are not invoked.

Reply:
starkscan won't recognize the contract as erc271 if i do not add these event in contract, i tested. It seems to be a bug and maybe the later cairo version will solve it.